transformers_cfg/parser.py

```diff
@@ -10,22 +13,163 @@
 END_OF_GRAMMAR_MARKER = 0xFFFF
 TO_BE_FILLED_MARKER = 0
 REF_RULE_MARKER = 1
-LITERAL_MARKER = 2
 
 
 ########################
 # EBNF Grammar Parsing #
 ########################
 
 
+class Codable(ABC):
+    def serialize(self) -> List[int]:
+        raise NotImplementedError()
+
+    @classmethod
+    def deserialize(cls, data: List[int]) -> "Codable":
+        raise NotImplementedError()
+
+
+@dataclass
+class GrammarElement(Codable):
+    def is_terminated(self) -> bool:
+        raise NotImplementedError()
+    
+
+@dataclass
+class TerminatedElement(GrammarElement):
```

Comment:
Maybe the the name `TerminalElement` would be more appropriate as terminal is the word used in grammar theory

Reply:
Just to see if I understand correctly, this `TerminalElement` is used to match a single char, isn't it ?

Reply:
The `TerminalElement` are all the elements that are not References. So it's just list of ranges. For a single char, it will be a list with one range: [(ord(char), ord(char)].